helix-view/src/watcher.rs

```diff
@@ -0,0 +1,341 @@
+use std::collections::{HashMap, HashSet};
+use std::fs;
+use std::path::{Path, PathBuf};
+
+use globset::{GlobBuilder, GlobMatcher};
+use notify::{
+    event::{CreateKind, ModifyKind, RemoveKind, RenameMode},
+    EventKind, RecommendedWatcher, RecursiveMode, Watcher,
+};
+use slotmap::SlotMap;
+use tokio::sync::{
+    mpsc::{self, UnboundedReceiver, UnboundedSender},
+    oneshot,
+};
+use tokio::task::JoinHandle;
+
+pub use notify::{event as notify_event, Event};
+
+// A list of directory names to ignore when found in the root of
+// a workspace. These are directories that would otherwise generate
+// a very large amount of events which could cause stuttering.
+// i.e. `rust-analyzer` calls `cargo check` on every write
+// which generates a large amount of fingerprint file changes in the
+// `target` directory.
+const IGNORE: &[&str] = &["target"];
```

Comment:
> ... ignoring those matching some default list of build/cache directories ala Rust's target directory

I think we can just use the .gitignore to filter files (or other .ignore files) to filter relevant files/folders.

They are used anyway already for the file-picker etc.

Reply:
I'm inclined to agree that `.gitignore` is the better approach now that I've thought about it some more. Originally I imagined that could lead to some surprising behaviour in edge cases depending on what the user's chosen to add to their `.gitignore`, but thinking about it now I can't really imagine many sane scenarios where something bad could happen.

For Rust and other languages that use such an integrated project/build system it should work well in the vast majority of cases, since of course `cargo new` automatically sets up a repo and gitignores `target`. Other languages I'm not so sure, but we can hope that the amount of cases where people would have an extremely active build directory that isn't present in `.gitignore` would be comparatively low.

This would also eliminate some edge cases like if a project is using a different name than `target` for its build directory.